token/wallet.tact

```diff
@@ -2,164 +2,116 @@
+        let required_input = transfer.forward_ton_amount + 
+            msg.fwd_fee * msg.fwd_count + 
+            2 * GAS_CONSUMPTION + 
+            STORAGE_MIN_CONSUMPTION;
+        verify(msg.amount > required_input, "not enough value");
+
+        self.balance -= transfer.amount;
+
+        Jetton.init(
+            msg.destination,
+            self.master,
+            self.wallet_code,
+            code: self.wallet_code
+        ).receive_tokens(InternalTransfer {
             query_id: msg.query_id,
             amount: msg.amount,
-            from: state.owner,
+            from: self.owner,
             response_address: msg.response_destination,
             forward_ton_amount: msg.forward_ton_amount,
             forward_payload: msg.forward_payload,
         })
-        }.serialize())
-        .send_to(to_wallet_address);
-}
-
-fn receive_tokens(
-    @mut state: WalletState, 
-    msg: InternalTransfer, 
-    sender: ContractAddress, 
-    my_ton_balance: Coins, 
-    fwd_fee: Coins, 
-    msg_value: Coins
-) {
-    verify(
-        sender === state.master || 
-            sender == WalletAddress.from_data(msg.from, state.master, state.wallet_code), 
-        "sender must be master or user wallet"
-    );
-
-    let ton_balance_before_msg = my_ton_balance - msg_value;
-    let storage_fee = STORAGE_MIN_CONSUMPTION - ton_balance_before_msg.min(STORAGE_MIN_CONSUMPTION);
-    let @mut msg_value = msg_value - (storage_fee + GAS_CONSUMPTION);
-
-    if (forward_ton_amount != 0) {
-        msg_value -= (forward_ton_amount + fwd_fee);
-        let payload = msg.forward_payload;
-        RequestBuilder.new()
-            .ihr_disabled()
-            .pay_transfer_fee_separately()
-            .money(forward_ton_amount)
-            .body(TransferNotification {
-                query_id: msg.query_id,
-                amount: msg.amount,
-                from: msg.from,
-                payload: payload,
-            })
-            .send_to(owner_address);
+        .send(
+            ihr_disabled: true,
+            can_be_bounced: true,
+            carry_message_input: true,
+        );
     }
 
-    if (!response_address.is_null() && msg_value > 0) {
-        RequestBuilder.new()
-            .ihr_disabled()
-            .ignore_action_phase_errors()
-            .money(msg_value)
-            .body(Excesses {
-                query_id: msg.query_id,
-            })
-            .send_to(response_address);
+    internal receive_tokens#178d4519(transfer: InternalTransfer) {
+        verify(
+            msg.sender === self.master || 
+                sender == WalletAddress.from_data(msg.from, self.master, self.wallet_code), 
+            "sender must be master or user wallet"
+        );
+
+        let ton_balance_before_msg = self.my_balance() - msg.amount;
+        let storage_fee = STORAGE_MIN_CONSUMPTION - ton_balance_before_msg.min(STORAGE_MIN_CONSUMPTION);
```

Comment:
```suggestion
        let storage_fee = STORAGE_MIN_CONSUMPTION - min(ton_balance_before_msg,STORAGE_MIN_CONSUMPTION);
```

i think `min`/`max` deserve top-level definitions for symmetric read, on par with basic arithmetic and we can simply add baked-in compiler support to avoid wasting time on making the generics more powerful to express `fn min(a: Int[A], b: Int[B])->Int[max(A,B)]` within Tact itself. We can always get back to this at the 2.0 stage.

Reply:
But `STORAGE_MIN_CONSUMPTION` and `ton_balance_before_msg` have type `Coins`.

Reply:
Coins is just VarUInt, VarUInt should be a wrapper over Int that behaves like it, but serializes differently. API should look sane: `max(a,b)`, we don't want so basic things to look like from Smalltalk land of crazy: `a.max(b)`.